fem/pfespace.cpp

```diff
@@ -3252,6 +3281,252 @@ void DeviceConformingProlongationOperator::MultTranspose(const Vector &x,
+         }
+         f_ind++;
+      }
+   }
+   MFEM_VERIFY(f_ind==nf, "Unexpected number of faces.");
+}
+
+void ParL2FaceRestriction::Mult(const Vector& x, Vector& y) const
+{
+   ParGridFunction x_gf;
+   x_gf.MakeRef(const_cast<ParFiniteElementSpace*>(&fes), const_cast<Vector&>(x),
+                0);
+   x_gf.ExchangeFaceNbrData();
+
+   // Assumes all elements have the same number of dofs
+   const int nd = dof;
+   const int vd = vdim;
+   const bool t = byvdim;
+   const int threshold = ndofs;
+
+   if (m==L2FaceValues::Double)
+   {
+      auto d_indices1 = indices1.Read();
+      auto d_indices2 = indices2.Read();
+      auto d_x = Reshape(x.Read(), t?vd:ndofs, t?ndofs:vd);
+      auto d_x_shared = Reshape(x_gf.FaceNbrData().Read(), t?vd:ndofs, t?ndofs:vd);
+      auto d_y = Reshape(y.Write(), nd, vd, 2, nf);
+      MFEM_FORALL(i, nfdofs,
+      {
+         const int dof = i % nd;
+         const int face = i / nd;
+         const int idx1 = d_indices1[i];
+         for (int c = 0; c < vd; ++c)
+         {
+            d_y(dof, c, 0, face) = d_x(t?c:idx1, t?idx1:c);
+         }
+         const int idx2 = d_indices2[i];
+         for (int c = 0; c < vd; ++c)
+         {
+            if (idx2>-1 && idx2<threshold) //interior face
+            {
+               d_y(dof, c, 1, face) = d_x(t?c:idx2, t?idx2:c);
+            }
+            else if (idx2>=threshold) //shared boundary
+            {
+               d_y(dof, c, 1, face) = d_x_shared(t?c:(idx2-threshold), t?(idx2-threshold):c);
+            }
+            else //true boundary
+            {
+               d_y(dof, c, 1, face) = 0.0;
+            }
+         }
+      });
+   }
+   else
+   {
+      auto d_indices1 = indices1.Read();
+      auto d_x = Reshape(x.Read(), t?vd:ndofs, t?ndofs:vd);
+      auto d_y = Reshape(y.Write(), nd, vd, nf);
+      MFEM_FORALL(i, nfdofs,
+      {
+         const int dof = i % nd;
+         const int face = i / nd;
+         const int idx1 = d_indices1[i];
+         for (int c = 0; c < vd; ++c)
+         {
+            d_y(dof, c, face) = d_x(t?c:idx1, t?idx1:c);
+         }
+      });
+   }
+}
+
+void ParL2FaceRestriction::MultTranspose(const Vector& x, Vector& y) const
+{
+   // Assumes all elements have the same number of dofs
+   const int nd = dof;
+   const int vd = vdim;
+   const bool t = byvdim;
+   const int threshold = ndofs;
+   if (m==L2FaceValues::Double)
+   {
+      auto d_indices1 = indices1.Read();
+      auto d_indices2 = indices2.Read();
+      auto d_x = Reshape(x.Read(), nd, vd, 2, nf);
+      auto d_y = Reshape(y.ReadWrite(), t?vd:ndofs, t?ndofs:vd);
+      MFEM_FORALL(i, nfdofs,
+      {
+         const int idx1 = d_indices1[i];
+         const int idx2 = d_indices2[i];
+         for (int c = 0; c < vd; ++c)
+         {
+            MFEM_ATOMIC_ADD(d_y(t?c:idx1,t?idx1:c), d_x(i % nd, c, 0, i / nd));
+            if (idx2>-1 && idx2<threshold) { MFEM_ATOMIC_ADD(d_y(t?c:idx2,t?idx2:c), d_x(i % nd, c, 1, i / nd)); }
```

Comment:
I see we are using atomics here. Could this be avoided easily?

Reply:
Not easily. But this can be avoided, using a quite different algorithm.

Reply:
It's easier than I thought, next thing on my TODO list.

Reply:
done.